### PR TITLE
Support hex/oct/binary representation of floating-point entries

### LINKED ISSUE #115: Fix float hexadecimal view
If you have a table entry whose data type is a float, clicking "View as Hexadecimal" does not change the value's representation.

### PR DESCRIPTION
**Pull request by malleoz**
This fixes #115

### CODE REVIEW

File: Source/Common/MemoryCommon.cpp [hunk 1]
```diff
@@ -490,11 +490,20 @@ std::string formatMemoryToString(const char* memory, const MemType type, const s
...
-    std::memcpy(&aFloat, memoryCopy, sizeof(float));
-    // With 9 digits of precision, it is possible to convert a float back and forth to its binary
-    // representation without any loss
-    ss << std::setprecision(9) << aFloat;
+    if (base == Common::MemBase::base_hexadecimal)
+    {
+      u32 unsignedWord = 0;
+      std::memcpy(&unsignedWord, memoryCopy, sizeof(u32));
+      ss << unsignedWord;
+    }
```
Comment (cristian64) [inline Source/Common/MemoryCommon.cpp:498]: While we are here, I think we can also support the other representations (octal and binary). Suggestion:

```cpp
    if (base != Common::MemBase::base_decimal)
    {
      u32 unsignedWord = 0;
      std::memcpy(&unsignedWord, memoryCopy, sizeof(u32));
      if (base == Common::MemBase::base_binary)
      {
        ss << std::bitset<sizeof(u32) * 8>(unsignedWord).to_string();
      }
      else
      {
        ss << unsignedWord;
      }
    }
    else
    {
      float aFloat = 0.0f;
      ...
    }
```

I think binary in particular can be useful.

(Similar change for the `double` type below.)

Comment (malleoz) [reply Source/Common/MemoryCommon.cpp:498]: Sure thing. I also noticed that I did not account for when the user types in a new float value in hex, so I will need to make appropriate changes for that too.

**Comment by malleoz** [comment]
I've updated the PR to address @cristian64's suggestion above. Testing looks good on my end for float/double reading/writing.

**Comment by cristian64** [comment]
@malleoz In case you've missed it, it seems the formatting is not compliant with the rules defined in the `.clang-format` file.

File: Source/Common/MemoryCommon.cpp [hunk 2]
```diff
@@ -264,34 +264,106 @@ char* formatStringToMemory(MemOperationReturnCode& returnCode, size_t& actualLen
...
-      return buffer;
+      u32 theWord = 0;
+      if (base == MemBase::base_binary)
+      {
+        unsigned long long input = 0;
+        try
+        {
+          input = std::bitset<sizeof(u32) * 8>(inputString).to_ullong();
+        }
+        catch (std::invalid_argument)
```
[the remark this answers is not in the document]
Comment (malleoz) [reply Source/Common/MemoryCommon.cpp]: Great catch. I've applied that change for both float and double cases.

File: Source/Common/MemoryCommon.cpp [hunk 3]
```diff
@@ -264,34 +264,106 @@ char* formatStringToMemory(MemOperationReturnCode& returnCode, size_t& actualLen
...
+          input = std::bitset<sizeof(u64) * 8>(inputString).to_ullong();
+        }
+        catch (std::invalid_argument)
+        {
+          delete[] buffer;
+          buffer = nullptr;
+          returnCode = MemOperationReturnCode::invalidInput;
+          return buffer;
+        }
+        theDoubleWord = static_cast<u32>(input);
```
[the remark this answers is not in the document]
Comment (malleoz) [reply Source/Common/MemoryCommon.cpp]: Yes, sorry! Fixed now.

**Comment by cristian64** [comment]
@malleoz It seems an unrelated commit has been added to the branch inadvertently.

Also, I think the one commit and PR title could be reworded to mention octal and binary representation too (not only hex anymore).

**Comment by malleoz** [comment]
Done and done. Sorry I had cherry-picked another commit to make my own executable for personal use, and forgot to remove when making changes!

**Review by cristian64 — APPROVED** [review]
Looks good. Tested on Linux.

**Review by dreamsyntax — APPROVED** [review]

**Comment by dreamsyntax** [comment]
Tested on Windows.

**Pull request merged**